netshare/generate/denormalize_fields.py

```diff
@@ -0,0 +1,138 @@
+            field_names.append(field.name)
+    return field_names
+
+
+def _denormalize_by_fields_list(
+    normalized_data: np.ndarray, fields_list: List[Field], is_session_key: bool
+) -> List[np.ndarray]:
+    """
+    This function executes field.denormalize for each of the given field.
+    """
+    denormalized_data = []
+    dim = 0
+
+    canonical_data_dir = get_canonical_data_dir()
+    df, _ = load_dataframe_chunks(canonical_data_dir)
+
+    for field in fields_list:
+        if is_session_key:
+            sub_data = normalized_data[:, dim : dim + field.get_output_dim()]
+        else:
+            sub_data = normalized_data[:, :, dim : dim + field.get_output_dim()]
+
+        sub_data = field.denormalize(sub_data)
+
+        # For session key, if shape looks like (n, ), change it to (n, 1) for consistency
+        if is_session_key and len(sub_data.shape) == 1:
+            sub_data = np.expand_dims(sub_data, axis=1)
+        # For timeseries, if shape looks like (i, j), change it to (i, j, 1) for consistency
+        if not is_session_key and len(sub_data.shape) == 2:
+            sub_data = np.expand_dims(sub_data, axis=2)
+        denormalized_data.append(sub_data)
+        dim += field.get_output_dim()
+    return denormalized_data
+
+
+def write_to_csv(
+    csv_folder: str,
+    session_key_fields: List[Field],
+    timeseries_fields: List[Field],
+    session_key: List[np.ndarray],
+    timeseries: List[np.ndarray],
+    data_gen_flag: np.ndarray,
+    filename: str,
+) -> None:
+    """
+    This function dumps the given data to the given directory as a csv format.
+    `data_gen_flag` is an indicator showing if the time series for this session
+    has ended in this time step.
+    """
+    os.makedirs(csv_folder, exist_ok=True)
+    csv_path = os.path.join(csv_folder, f"data_{filename}_{random.random()}.csv")
+    # change session key shape to #session * #attributes
+    session_key = np.concatenate(session_key, axis=1)
+    # change timeseries shape to #session * #time_steps * #features
+    timeseries = np.concatenate(timeseries, axis=2)
+
+    with open(csv_path, "w") as f:
+        writer = csv.writer(f)
+        session_titles = _get_fields_names(session_key_fields)
+        timeseries_titles = _get_fields_names(timeseries_fields)
+        writer.writerow(session_titles + timeseries_titles)
+
+        for (
+            data_gen_per_session,
+            session_data_per_session,
+            timeseries_per_session,
+        ) in zip(data_gen_flag, session_key, timeseries):
+            # this if is here in parallel to the if in `reduce_samples`. It supports old flows.
+            session_data_per_session = session_data_per_session.tolist()
+            for j in range(data_gen_per_session.shape[0]):
+                if data_gen_per_session[j] == 1.0:
+                    timeseries_data = timeseries_per_session[j].tolist()
+                    writer.writerow(session_data_per_session + timeseries_data)
```

Comment:
Not sure what the comment (line 94) means here -- I thought line 97-99 are compulsory by default since not all flows are the same length?

Reply:
Oh, nothing related to the bug and I think the code itself makes sense -- anyway we need `data_gen_flag` to extract the "valid" timeseries. I was just confused about the comments. Thanks for the explanation!